groups/bsl/bslstl/bslstl_optional.t.cpp

```diff
@@ -0,0 +1,1032 @@
+    //:   comprehensive testing of 'bsl::optional'.  This class manages the
+    //:   buffer that stores the optional's value type object to ensure access
+    //:   to the value type object is 'const' if the value type is 'const'.
+    //:   The class also provides an interface for creating the value type
+    //:   object inside the buffer, which will be comprehensively tested
+    //:   through the interface of 'bsl::optional'.
+    //
+    // Plan:
+    //: 1 Create an empty 'Optional_DataImp' object
+    //: 2 Test that 'has_value' returns 'false'
+    //: 3 Emplace a value into the 'Optional_DataImp' object and check that
+    //:   'value()' method returns the emplaced value
+    //: 4 Check that the 'value' method returns a const object if 'TYPE' is
+    //:   const qualified
+    //: 5 Call 'reset' and verify that 'has_value' returns 'false'
+    //
+    // Testing:
+    //   'Optional_DataImp' class
+    // --------------------------------------------------------------------
+
+    if (verbose)
+        printf("\n'Optional_DataImp' TEST"
+               "\n=======================\n");
+
+    bslma::TestAllocator da("default", veryVeryVeryVerbose);
+
+    BloombergLP::bslstl::Optional_DataImp<TYPE> X;
+    ASSERT(!X.hasValue());
+
+    TYPE val = TYPE();
+    X.emplace(&da, val);
+    ASSERT(X.hasValue());
+    ASSERT(X.value() == val);
+    ASSERT(checkAllocator(X, &da));
+
+    ASSERT(isConstPtr(&val) == isConstPtr(&X.value()));
+
+    X.reset();
+    ASSERT(!X.hasValue());
+}
+//=============================================================================
+//                              MAIN PROGRAM
+//-----------------------------------------------------------------------------
+
+int main(int argc, char **argv)
+{
+    const int test      = argc > 1 ? atoi(argv[1]) : 0;
+    verbose             = argc > 2;
+    veryVerbose         = argc > 3;
+    veryVeryVerbose     = argc > 4;
+    veryVeryVeryVerbose = argc > 5;
+
+    printf("TEST  %s CASE %d \n", __FILE__, test);
+
+    switch (test) {
+      case 0:
+      case 2: {
+        // --------------------------------------------------------------------
+        // 'Optional_DataImp' TEST
+        // --------------------------------------------------------------------
+
+        if (verbose)
+            printf("\n"
+                   "'Optional_DataImp' TEST"
+                   "\n"
+                   "======================="
+                   "\n");
+
+        TestDriver<int>::bslstl_optional_test2();
+        TestDriver<MyClass1>::bslstl_optional_test2();
+        TestDriver<MyClass2>::bslstl_optional_test2();
+        TestDriver<MyClass2a>::bslstl_optional_test2();
+        TestDriver<const int>::bslstl_optional_test2();
+        TestDriver<const MyClass1>::bslstl_optional_test2();
+        TestDriver<const MyClass2>::bslstl_optional_test2();
+        TestDriver<const MyClass2a>::bslstl_optional_test2();
```

Comment:
Please look at `BSLTF_TEMPLATETESTFACILITY_RUN_EACH_TYPE` and how it is used in our test drivers - might be useful, including things like `BSLTF_TEMPLATETESTFACILITY_TEST_TYPES_REGULAR`, etc.

Reply:
noted. I'm in the process of rewriting the tests based on an earlier suggestion from Mike (the test setup is currently based on bslmla_constructionutil tests). I think it would make sense to review parts of bsl::optional (constructors/assignment operators/remaining methods and free functions) before reviewing the tests. That way, I can work on tests while you review the bsl::optional code. WDYT ?